fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContextTest.java

```diff
@@ -0,0 +1,280 @@
+                    .limit(10)
+                    .collect(Collectors.toList());
+            List<Long> readVersions = AsyncUtil.getAll(futures).get();
+            long firstReadVersion = readVersions.get(0);
+            for (long readVersion : readVersions) {
+                assertEquals(firstReadVersion, readVersion);
+            }
+            assertEquals(1, timer.getCount(FDBStoreTimer.Events.GET_READ_VERSION));
+        }
+    }
+
+    @Test
+    public void getReadVersionTimingWithInjectedLatency() {
+        final FDBStoreTimer timer = new FDBStoreTimer();
+        FDBDatabaseFactory factory = fdb.getFactory();
+        Function<FDBLatencySource, Long> oldLatencyInjector = factory.getLatencyInjector();
+        factory.setLatencyInjector(latencySource -> {
+            if (latencySource.equals(FDBLatencySource.GET_READ_VERSION)) {
+                return 50L;
+            } else {
+                return 10L;
+            }
+        });
+        factory.clear();
+        fdb = factory.getDatabase(fdb.getClusterFile());
+        try {
+            try (FDBRecordContext context = fdb.openContext(null, timer)) {
+                context.getReadVersion();
+            }
+            assertEquals(1, timer.getCount(FDBStoreTimer.Waits.WAIT_GET_READ_VERSION));
+            assertEquals(1, timer.getCount(FDBStoreTimer.Events.GET_READ_VERSION));
+            assertEquals(1, timer.getCount(FDBStoreTimer.Events.INJECTED_GET_READ_VERSION_LATENCY));
+
+            long waitNanos = timer.getTimeNanos(FDBStoreTimer.Waits.WAIT_GET_READ_VERSION);
+            long grvNanos = timer.getTimeNanos(FDBStoreTimer.Events.GET_READ_VERSION);
+            long injectedNanos = timer.getTimeNanos(FDBStoreTimer.Events.INJECTED_GET_READ_VERSION_LATENCY);
+
+            // Validate that the grvNanos includes the injected nanos. It is very likely that GRV is higher than
+            // the injected nanos, but theoretically could be different
+            assertThat(injectedNanos, lessThanOrEqualTo(grvNanos));
+            assertThat(waitNanos, greaterThan(0L));
+
+        } finally {
+            factory.setLatencyInjector(oldLatencyInjector);
+            factory.clear();
+        }
+    }
+
+    @ParameterizedTest(name = "closeWithOutstandingGetReadVersion [inject latency = {0}]")
+    @EnumSource(TestHelpers.BooleanEnum.class)
+    public void closeWithOutstandingGetReadVersion(@Nonnull TestHelpers.BooleanEnum injectLatency) throws InterruptedException, ExecutionException {
+        FDBDatabaseFactory factory = fdb.getFactory();
+        Function<FDBLatencySource, Long> oldLatencyInjector = factory.getLatencyInjector();
+        if (injectLatency.toBoolean()) {
+            factory.setLatencyInjector(latencySource -> 5L);
+        } else {
+            factory.setLatencyInjector(latencySource -> 0L);
+        }
+        factory.clear();
+        fdb = factory.getDatabase(fdb.getClusterFile());
```

Comment:
if `factory.getDatabase(fdb.getClusterFile())` fails, the latency injector will not be set back. I'm not sure if `factory.clear() has the potential to throw an exception, but you probably ant to extend the try block to at least include this.

Reply:
Sure. Moved everything that I could.